Mathlib/Analysis/NormedSpace/ProdLp.lean

```diff
@@ -0,0 +1,743 @@
+  induction p using ENNReal.recTopCoe generalizing hp with
+  | top =>
+    simp [prod_nnnorm_eq_sup]
+  | coe p =>
+    have hp0 : (p : ℝ) ≠ 0 := by
+      exact_mod_cast (zero_lt_one.trans_le <| Fact.out (p := 1 ≤ (p : ℝ≥0∞))).ne'
+    simp [prod_nnnorm_eq_add, NNReal.zero_rpow hp0, ← NNReal.rpow_mul, mul_inv_cancel hp0]
+
+@[simp]
+theorem nnnorm_equiv_symm_snd [hp : Fact (1 ≤ p)] (y : β) :
+    ‖(WithLp.equiv p (α × β)).symm (0, y)‖₊ = ‖y‖₊ := by
+  induction p using ENNReal.recTopCoe generalizing hp with
+  | top =>
+    simp [prod_nnnorm_eq_sup]
+  | coe p =>
+    have hp0 : (p : ℝ) ≠ 0 := by
+      exact_mod_cast (zero_lt_one.trans_le <| Fact.out (p := 1 ≤ (p : ℝ≥0∞))).ne'
+    simp [prod_nnnorm_eq_add, NNReal.zero_rpow hp0, ← NNReal.rpow_mul, mul_inv_cancel hp0]
+
+@[simp]
+theorem norm_equiv_symm_fst (x : α) : ‖(WithLp.equiv p (α × β)).symm (x, 0)‖ = ‖x‖ :=
+  congr_arg ((↑) : ℝ≥0 → ℝ) <| nnnorm_equiv_symm_fst p α β x
+
+@[simp]
+theorem norm_equiv_symm_snd (y : β) : ‖(WithLp.equiv p (α × β)).symm (0, y)‖ = ‖y‖ :=
+  congr_arg ((↑) : ℝ≥0 → ℝ) <| nnnorm_equiv_symm_snd p α β y
+
+@[simp]
+theorem nndist_equiv_symm_single_fst (x₁ x₂ : α) :
+    nndist ((WithLp.equiv p (α × β)).symm (x₁, 0)) ((WithLp.equiv p (α × β)).symm (x₂, 0)) =
+      nndist x₁ x₂ := by
+  rw [nndist_eq_nnnorm, nndist_eq_nnnorm, ← WithLp.equiv_symm_sub, Prod.mk_sub_mk, sub_zero,
+    nnnorm_equiv_symm_fst]
+
+@[simp]
+theorem nndist_equiv_symm_single_snd (y₁ y₂ : β) :
+    nndist ((WithLp.equiv p (α × β)).symm (0, y₁)) ((WithLp.equiv p (α × β)).symm (0, y₂)) =
+      nndist y₁ y₂ := by
+  rw [nndist_eq_nnnorm, nndist_eq_nnnorm, ← WithLp.equiv_symm_sub, Prod.mk_sub_mk, sub_zero,
+    nnnorm_equiv_symm_snd]
+
+@[simp]
+theorem dist_equiv_symm_single_fst (x₁ x₂ : α) :
+    dist ((WithLp.equiv p (α × β)).symm (x₁, 0)) ((WithLp.equiv p (α × β)).symm (x₂, 0)) =
+      dist x₁ x₂ :=
+  congr_arg ((↑) : ℝ≥0 → ℝ) <| nndist_equiv_symm_single_fst p α β x₁ x₂
+
+@[simp]
+theorem dist_equiv_symm_single_snd (y₁ y₂ : β) :
+    dist ((WithLp.equiv p (α × β)).symm (0, y₁)) ((WithLp.equiv p (α × β)).symm (0, y₂)) =
+      dist y₁ y₂ :=
+  congr_arg ((↑) : ℝ≥0 → ℝ) <| nndist_equiv_symm_single_snd p α β y₁ y₂
+
+@[simp]
+theorem edist_equiv_symm_single_fst (x₁ x₂ : α) :
+    edist ((WithLp.equiv p (α × β)).symm (x₁, 0)) ((WithLp.equiv p (α × β)).symm (x₂, 0)) =
+      edist x₁ x₂ := by
+  simp only [edist_nndist, nndist_equiv_symm_single_fst p α β x₁ x₂]
+
+@[simp]
+theorem edist_equiv_symm_single_snd (y₁ y₂ : β) :
+    edist ((WithLp.equiv p (α × β)).symm (0, y₁)) ((WithLp.equiv p (α × β)).symm (0, y₂)) =
+      edist y₁ y₂ := by
+  simp only [edist_nndist, nndist_equiv_symm_single_snd p α β y₁ y₂]
+
+end Single
+
+variable (𝕜 p α β)
+
+/-- `WithLp.equiv` as a continuous linear equivalence. -/
+@[simps! (config := { fullyApplied := false }) apply symm_apply]
+protected def prodContinuousLinearEquiv : WithLp p (α × β) ≃L[𝕜] α × β where
```

Comment:
I think this also works with weaker type classes. Presumably Pseudo(E? probbably not)Metric

Reply:
in fact you only need the topological space instances, which are for free